src/v/kafka/server/handlers/configs/config_response_utils.cc

```diff
@@ -312,6 +312,36 @@ override_if_not_default(const std::optional<T>& override, const T& def) {
     }
 }
 
+// The cases in which the override is considered equivalent to the default are:
+// 1. `tristate` is empty (`std::nullopt`) and the optional default is also
+// empty (`std::nullopt`)
+// 2. `tristate` is disabled (`disable_tristate`) and the optional default is
+// empty (`std::nullopt`).
+// 3. `tristate` holds a value and the optional default holds the same value.
+template<typename T>
+tristate<T> override_if_not_default(
+  const tristate<T>& overrides, const std::optional<T>& def) {
+    // Overrides is {disabled}, cluster default is {value}
+    if (overrides.is_disabled() && def.has_value()) {
```

Comment:
why does it matter if def.has_value here?

Reply:
For the two cases you have commented on (`// Overrides is {disabled}, cluster default is {value}`, `// Overrides is {std::nullopt}, cluster default is {value}`), note that the following case in which `(overrides.is_empty() || overrides.is_disabled()) && !def.has_value()` is excluded.

I.e, I'm treating the override and default as "equivalent" when the `tristate` is either disabled or empty, and the optional is in the `std::nullopt` state.

This makes a lot of logical sense for `overrides.is_empty() && !def.has_value()`, but I recognize `overrides.is_disabled() && !def.has_value()` is a bit more of a stretch. However, in the case of `delete.retention.ms`, its default state is often `disabled`, and the corresponding cluster property `tombstone_retention_ms`'s default state is often `std::nullopt`. This, IMO, provides motivation for considering this state in `override_if_not_default()` in the way it is currently written.

Reply:
> override is considered equivalent to the default are

do you think we could add to the comment a truth table for all 6 cases (or maybe it's 9 if we count the cases where the default has a value and its actually equal to the tristate rather than just semantics equal). then we could use the retention.ms and truth table as an example for override_if_not_default usage.

Reply:
Added truth table in a code comment